File: LogisimPlus/LogisimPlus/logisim/analyze/gui/OutputSelector.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// To produce customized conversions, purchase a Premium Edition license:
// https://www.tangiblesoftwaresolutions.com/product-details/java-to-csharp-converter.html
// ====================================================================================================

/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

namespace logisim.analyze.gui
{


	using AnalyzerModel = logisim.analyze.model.AnalyzerModel;
	using VariableList = logisim.analyze.model.VariableList;
	using VariableListEvent = logisim.analyze.model.VariableListEvent;
	using VariableListListener = logisim.analyze.model.VariableListListener;

	internal class OutputSelector
	{
		private class Model : AbstractListModel<string>, ComboBoxModel<string>, VariableListListener
		{
			private readonly OutputSelector outerInstance;

			public Model(OutputSelector outerInstance)
			{
				this.outerInstance = outerInstance;
			}

			internal string selected;

			public virtual object SelectedItem
			{
				set
				{
					selected = (string)value;
				}
				get
				{
					return selected;
				}
			}


			public virtual int Size
			{
				get
				{
					return outerInstance.source.size();
				}
			}

			public virtual string getElementAt(int index)
			{
				return outerInstance.source.get(index);
			}

			public virtual void listChanged(VariableListEvent @event)
			{
				int index;
				string variable;
				object selection;
				switch (@event.Type)
				{
				case VariableListEvent.ALL_REPLACED:
					outerInstance.computePrototypeValue();
					fireContentsChanged(this, 0, Size);
					if (outerInstance.source.Empty)
					{
						outerInstance.select.setSelectedItem(null);
					}
					else
					{
						outerInstance.select.setSelectedItem(outerInstance.source.get(0));
					}
					break;
				case VariableListEvent.ADD:
					variable = @event.Variable;
					if (string.ReferenceEquals(outerInstance.prototypeValue, null) || variable.Length > outerInstance.prototypeValue.Length)
					{
						outerInstance.computePrototypeValue();
					}

					index = outerInstance.source.indexOf(variable);
					fireIntervalAdded(this, index, index);
					if (outerInstance.select.getSelectedItem() == null)
					{
						outerInstance.select.setSelectedItem(variable);
					}
					break;
				case VariableListEvent.REMOVE:
					variable = @event.Variable;
					if (variable.Equals(outerInstance.prototypeValue))
					{
						outerInstance.computePrototypeValue();
					}
					index = ((int?) @event.Data).Value;
					fireIntervalRemoved(this, index, index);
					selection = outerInstance.select.getSelectedItem();
					if (selection != null && selection.Equals(variable))
					{
						selection = outerInstance.source.Empty ? null : outerInstance.source.get(0);
						outerInstance.select.setSelectedItem(selection);
					}
					break;
				case VariableListEvent.MOVE:
					fireContentsChanged(this, 0, Size);
					break;
				case VariableListEvent.REPLACE:
					variable = @event.Variable;
					if (variable.Equals(outerInstance.prototypeValue))
					{
						outerInstance.computePrototypeValue();
					}
					index = ((int?) @event.Data).Value;
					fireContentsChanged(this, index, index);
					selection = outerInstance.select.getSelectedItem();
					if (selection != null && selection.Equals(variable))
					{
						outerInstance.select.setSelectedItem(@event.Source.get(index));
					}
					break;
				}
			}
		}

		private VariableList source;
		private JLabel label = new JLabel();
		private JComboBox<string> select = new JComboBox<string>();
		private string prototypeValue = null;

		public OutputSelector(AnalyzerModel model)
		{
			this.source = model.Outputs;

			Model listModel = new Model(this);
			select.setModel(listModel);
			source.addVariableListListener(listModel);
		}

		public virtual JPanel createPanel()
		{
			JPanel ret = new JPanel();
			ret.add(label);
			ret.add(select);
			return ret;
		}

		public virtual JLabel Label
		{
			get
			{
				return label;
			}
		}

		public virtual JComboBox<string> ComboBox
		{
			get
			{
				return select;
			}
		}

		internal virtual void localeChanged()
		{
			label.setText(Strings.get("outputSelectLabel"));
		}

		public virtual void addItemListener(ItemListener l)
		{
			select.addItemListener(l);
		}

		public virtual void removeItemListener(ItemListener l)
		{
			select.removeItemListener(l);
		}

		public virtual string SelectedOutput
		{
			get
			{
				string value = (string) select.getSelectedItem();
				if (!string.ReferenceEquals(value, null) && !source.contains(value))
				{
					if (source.Empty)
					{
						value = null;
					}
					else
					{
						value = source.get(0);
					}
					select.setSelectedItem(value);
				}
				return value;
			}
		}

		private void computePrototypeValue()
		{
			string newValue;
			if (source.Empty)
			{
				newValue = "xx";
			}
			else
			{
				newValue = "xx";
				for (int i = 0, n = source.size(); i < n; i++)
				{
					string candidate = source.get(i);
					if (candidate.Length > newValue.Length)
					{
						newValue = candidate;
					}
				}
			}
			if (string.ReferenceEquals(prototypeValue, null) || newValue.Length != prototypeValue.Length)
			{
				prototypeValue = newValue;
				select.setPrototypeDisplayValue(prototypeValue + "xx");
				select.revalidate();
			}
		}
	}

}

File: logisim/app/src/main/java/logisim/analyze/gui/OutputSelector.java
/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

package logisim.analyze.gui;

import java.awt.event.ItemListener;

import javax.swing.AbstractListModel;
import javax.swing.ComboBoxModel;
import javax.swing.JComboBox;
import javax.swing.JLabel;
import javax.swing.JPanel;

import logisim.analyze.model.AnalyzerModel;
import logisim.analyze.model.VariableList;
import logisim.analyze.model.VariableListEvent;
import logisim.analyze.model.VariableListListener;

class OutputSelector {
	private class Model extends AbstractListModel<String> implements ComboBoxModel<String>, VariableListListener {
		private String selected;

		public void setSelectedItem(Object value) {
			selected = (String)value;
		}

		public Object getSelectedItem() {
			return selected;
		}

		public int getSize() {
			return source.size();
		}

		public String getElementAt(int index) {
			return source.get(index);
		}

		public void listChanged(VariableListEvent event) {
			int index;
			String variable;
			Object selection;
			switch (event.getType()) {
			case VariableListEvent.ALL_REPLACED:
				computePrototypeValue();
				fireContentsChanged(this, 0, getSize());
				select.setSelectedItem(source.isEmpty() ? null : source.get(0));
				break;
			case VariableListEvent.ADD:
				variable = event.getVariable();
				if (prototypeValue == null || variable.length() > prototypeValue.length())
					computePrototypeValue();

				index = source.indexOf(variable);
				fireIntervalAdded(this, index, index);
				if (select.getSelectedItem() == null)
					select.setSelectedItem(variable);
				break;
			case VariableListEvent.REMOVE:
				variable = event.getVariable();
				if (variable.equals(prototypeValue))
					computePrototypeValue();
				index = (Integer) event.getData();
				fireIntervalRemoved(this, index, index);
				selection = select.getSelectedItem();
				if (variable.equals(selection)) {
					selection = source.isEmpty() ? null : source.get(0);
					select.setSelectedItem(selection);
				}
				break;
			case VariableListEvent.MOVE:
				fireContentsChanged(this, 0, getSize());
				break;
			case VariableListEvent.REPLACE:
				variable = event.getVariable();
				if (variable.equals(prototypeValue))
					computePrototypeValue();
				index = (Integer) event.getData();
				fireContentsChanged(this, index, index);
				selection = select.getSelectedItem();
				if (variable.equals(selection))
					select.setSelectedItem(event.getSource().get(index));
				break;
			}
		}
	}

	private VariableList source;
	private JLabel label = new JLabel();
	private JComboBox<String> select = new JComboBox<>();
	private String prototypeValue;

	public OutputSelector(AnalyzerModel model) {
		source = model.getOutputs();

		Model listModel = new Model();
		select.setModel(listModel);
		source.addVariableListListener(listModel);
	}

	public JPanel createPanel() {
		JPanel ret = new JPanel();
		ret.add(label);
		ret.add(select);
		return ret;
	}

	public JLabel getLabel() {
		return label;
	}

	public JComboBox<String> getComboBox() {
		return select;
	}

	void localeChanged() {
		label.setText(Strings.get("outputSelectLabel"));
	}

	public void addItemListener(ItemListener l) {
		select.addItemListener(l);
	}

	public void removeItemListener(ItemListener l) {
		select.removeItemListener(l);
	}

	public String getSelectedOutput() {
		String value = (String) select.getSelectedItem();
		if (value == null || source.contains(value))
			return value;
		value = source.isEmpty() ? null : source.get(0);
		select.setSelectedItem(value);
		return value;
	}

	private void computePrototypeValue() {
		String newValue = source.stream()
				.max((a, b) -> b.length() - a.length())
				.orElse("xx");
		if (prototypeValue == null || newValue.length() != prototypeValue.length()) {
			prototypeValue = newValue;
			select.setPrototypeDisplayValue(prototypeValue + "xx");
			select.revalidate();
		}
	}
}
